### PR TITLE
PYTHON-3464 Add FaaS platform to handshake metadata

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Implements:
- https://jira.mongodb.org/browse/PYTHON-3464
- and part of https://jira.mongodb.org/browse/PYTHON-2052

Helpful links:
- https://github.com/mongodb/specifications/blob/5112bcc/source/mongodb-handshake/handshake.rst#limitations
- https://github.com/mongodb/specifications/blob/5112bcc/source/mongodb-handshake/handshake.rst#client-env

### CODE REVIEW

**Review by blink1073 — APPROVED** [review]
LGTM!

**Comment by blink1073** [comment]
The docs failure is unrelated and being worked on by RTD: https://github.com/readthedocs/readthedocs.org/issues/10290#issuecomment-1535446266

**Comment by ShaneHarvey** [comment]
Something funky is going on on windows:
```
 [2023/05/04 20:59:48.631] FAIL [0.004s]: test_metadata (test_client.ClientUnitTest)
 [2023/05/04 20:59:48.631] ----------------------------------------------------------------------
 [2023/05/04 20:59:48.631] Traceback (most recent call last):
 [2023/05/04 20:59:48.631]   File "C:\data\mci\8a18dfd850b927c5f947ec33058782e2\src\test\test_client.py", line 312, in test_metadata
 [2023/05/04 20:59:48.631]     self.assertEqual(options.pool_options.metadata, metadata)
 [2023/05/04 20:59:48.631] AssertionError: SON([[232 chars]cation', {'name': 'foobar'}), ('env', {'name': 'aws.lambda'})]) != SON([[232 chars]cation', {'name': 'foobar'})])
 [2023/05/04 20:59:48.631] ======================================================================
 [2023/05/04 20:59:48.631] FAIL [0.115s]: test_handshake_02_azure (test_client.TestExhaustCursor)
 [2023/05/04 20:59:48.631] ----------------------------------------------------------------------
 [2023/05/04 20:59:48.631] Traceback (most recent call last):
 [2023/05/04 20:59:48.631]   File "C:\data\mci\8a18dfd850b927c5f947ec33058782e2\src\test\test_client.py", line 1893, in test_handshake_02_azure
 [2023/05/04 20:59:48.631]     self._test_handshake({"FUNCTIONS_WORKER_RUNTIME": "python"}, {"name": "azure.func"})
 [2023/05/04 20:59:48.631]   File "C:\data\mci\8a18dfd850b927c5f947ec33058782e2\src\test\test_client.py", line 1880, in _test_handshake
 [2023/05/04 20:59:48.631]     self.assertEqual(options.pool_options.metadata, metadata)
 [2023/05/04 20:59:48.631] AssertionError: SON([[195 chars]', 'CPython 3.9.13.final.0')]) != SON([[195 chars]', 'CPython 3.9.13.final.0'), ('env', {'name': 'azure.func'})])
 [2023/05/04 20:59:48.631] ======================================================================
 [2023/05/04 20:59:48.631] FAIL [0.110s]: test_handshake_03_gcp (test_client.TestExhaustCursor)
 [2023/05/04 20:59:48.631] ----------------------------------------------------------------------
 [2023/05/04 20:59:48.631] Traceback (most recent call last):
 [2023/05/04 20:59:48.631]   File "C:\data\mci\8a18dfd850b927c5f947ec33058782e2\src\test\test_client.py", line 1896, in test_handshake_03_gcp
 [2023/05/04 20:59:48.632]     self._test_handshake(
 [2023/05/04 20:59:48.632]   File "C:\data\mci\8a18dfd850b927c5f947ec33058782e2\src\test\test_client.py", line 1880, in _test_handshake
 [2023/05/04 20:59:48.632]     self.assertEqual(options.pool_options.metadata, metadata)
 [2023/05/04 20:59:48.632] AssertionError: SON([[218 chars]l.0')]) != SON([[218 chars]l.0'), ('env', {'name': 'gcp.func', 'region': [50 chars]0})])
 [2023/05/04 20:59:48.632] ======================================================================
 [2023/05/04 20:59:48.632] FAIL [0.104s]: test_handshake_04_vercel (test_client.TestExhaustCursor)
 [2023/05/04 20:59:48.632] ----------------------------------------------------------------------
 [2023/05/04 20:59:48.632] Traceback (most recent call last):
 [2023/05/04 20:59:48.632]   File "C:\data\mci\8a18dfd850b927c5f947ec33058782e2\src\test\test_client.py", line 1907, in test_handshake_04_vercel
 [2023/05/04 20:59:48.632]     self._test_handshake(
 [2023/05/04 20:59:48.632]   File "C:\data\mci\8a18dfd850b927c5f947ec33058782e2\src\test\test_client.py", line 1880, in _test_handshake
 [2023/05/04 20:59:48.632]     self.assertEqual(options.pool_options.metadata, metadata)
 [2023/05/04 20:59:48.632] AssertionError: SON([[209 chars]9.13.final.0')]) != SON([[209 chars]9.13.final.0'), ('env', {'name': 'vercel', 'region': 'cdg1'})])
```
https://spruce.mongodb.com/task/mongo_python_driver_tests_windows_python_version__platform~windows_64_vsMulti_small_auth_ssl~auth_ssl_python_version_windows~3.9_test_rapid_replica_set_patch_d340710e3d488f3906d0e3fa4d4f25e56779fc3a_645416552fbabe7cdbad46d7_23_05_04_20_32_22/logs?execution=0&sortBy=STATUS&sortDir=ASC

File: pymongo/pool.py [hunk 1]
```diff
@@ -231,6 +232,108 @@ def _set_keepalive_times(sock):
     )
 
 
+def _is_lambda() -> bool:
+    if os.getenv("AWS_LAMBDA_RUNTIME_API"):
+        return True
+    env = os.getenv("AWS_EXECUTION_ENV")
+    if env:
+        return env.startswith("AWS_Lambda_")
```
Comment (ShaneHarvey) [inline pymongo/pool.py:240]: This change was needed to fix the issue described in https://jira.mongodb.org/browse/DRIVERS-2623

**Review by blink1073 — APPROVED** [review]

**Pull request merged**